### PR TITLE
Release: W14

### PR DESCRIPTION
**Pull request by C0DK**
### Changelog
 - Added explaination of languages used in HCL compiler
- Added explaination regarding high order functions
 - Added symbol table

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/3.Theory/LanguageSpecification.tex [hunk 1]
```diff
@@ -49,21 +49,27 @@ \section{Language Specification}
...
 A more explicit declaration of the same function can be seen in the snippet below:
 \begin{lstlisting}
-func<num,num,none> addNumbers = (num a, num b) -> none {
-	a + b print
+func[num,num,none] addNumbers = (num a, num b): num {
+	return a + b
 }
 \end{lstlisting}
 
+The return statement is noticeably missing in the first funciton.
```
Comment (nillas12) [inline report/3.Theory/LanguageSpecification.tex]: Fix Funciton

File: report/3.Theory/LanguageSpecification.tex [hunk 2]
```diff
@@ -126,6 +132,29 @@ \section{Language Specification}
 Not having operators for arithmetic operations also means that there are no native operator precedence in the language.
 This can be amended by structuring calculations in a left-to-right manner, or by encapsulating operations with parenthesis. 
 
+\textbf{High order functions}\\
+As mentioned in section \ref{subsec:hofhcl}, the language will include high-order functions. 
+To declare that a function will accept other functions, that the developer should simply declare a parameter of type "func" with the expected input and output types from the function.
```
Comment (nillas12) [inline report/3.Theory/LanguageSpecification.tex]: Remove that after the comma

File: report/3.Theory/LanguageSpecification.tex [hunk 3]
```diff
@@ -135,31 +164,26 @@ \section{Language Specification}
...
-
-\begin{lstlisting}
-1 to 10 sum print
-\end{lstlisting}
-
-Obviously the sum function accepts a list that should be summed. 
-However, if the sum function could potentially also accept a modifier function, that were to alter the result of the sum function, it would be impossible for the compiler to determine whether print was that modifier function, or the sum should be passed as a parameter to the print function.
+Function overloading in HCL is limited in that functions can not be overloaded with same parameters but different return types. 
+Functions can only be overloaded to return different types, if the types of the parameters vary.
+Furthermore overloading with different amounts of parameters are also not possible in HCL.
```
Comment (nillas12) [inline report/3.Theory/LanguageSpecification.tex]: Change "are also" to "is"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 4]
```diff
@@ -0,0 +1,38 @@
+\section{The languages of the compiler}
+When creating a compiler there are three languages of relevance, as seen in the tombstone diagram\ref{fig:TStoneExample}.
+
+\begin{figure}[H]
+	\centering
+	\includegraphics[width=\textwidth/2+\textwidth/4]{4.Solution/images/T-diagram.png}
+	\caption{
+		Example of Tombstone diagram showcasing an example of three languages for a given compiler.\cite{TStoneWiki}
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: Fix Cite after dot.

File: report/4.Solution/LanuageOfCompiler.tex [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	\label{fig:TStoneExample}
+\end{figure}
+In the case of the HCL compiler the three languages are as follows:
+
+\begin{align*}
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: Remember to fix weird table (Something you, Casper, said)

File: report/4.Solution/LanuageOfCompiler.tex [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+input \to & HCL \\
+output \to & C++/C \\
+compiler \to & Kotlin \\
+\end{align*}
+
+\subsection{Language of the compiler}
+The compiler was written in Kotlin\cite{KotlinWebsite}, a new programming language that compiles to both the JVM, javascript and LLVM. 
+The compiler is compiled to JVM because this grants access to all of the java ecosystem libraries, which in turn makes the JVM the most used target of kotlin.
+This choice to use kotlin was based on a number of factors.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: Change "This" (first word) to "The"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+The group wanted to experiment with using a language they had no prior experience with for the duration of the project, but because of time limitations, a new language had to adhere to already known paradigms.
+The group knew C\# and a bit of Java, however as the HCL language has a lot of functional features, it was deemed important that the group became comfortable with a language that, at the very least was multi paradigm.
+Kotlin is inspired by both C\# and java and have a fairly similar syntax to that of those, which made it easier to learn from the perspective of the group.
+However it is still possible to write methods that is utilizes a functional domain.
+
+This grants the developer a large amount of possibilities, while still being relatively easy to learn.
+A disadvantage could be that low-level languages might have faster executing time, however as HCL is built with the Arduino architecture in mind, it is not possible to create vast programs, as the architecture has some hardware limitations, so execution time of the compiler will never really affect the end user, as the difference should not be noticeable.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: This sentence is perhaps a bit longer than it should have been, perhaps consider splitting it (OPTIONAL)

File: report/3.Theory/LanguageSpecification.tex (hunk 1 above)
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: funciton -> function

File: report/3.Theory/LanguageSpecification.tex [hunk 8]
```diff
@@ -49,21 +49,27 @@ \section{Language Specification}
...
-func<num,num,none> addNumbers = (num a, num b) -> none {
-	a + b print
+func[num,num,none] addNumbers = (num a, num b): num {
+	return a + b
 }
 \end{lstlisting}
 
+The return statement is noticeably missing in the first funciton.
+"return" is optional if the function body only contains a single expression. 
+This makes sense when using using lambdas, where return statements are often implicitly written.
```
Comment (Abildgaard96) [inline report/3.Theory/LanguageSpecification.tex]: using using
too many using

File: report/4.Solution/LanuageOfCompiler.tex [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+\subsection{Output language of the compiler}
+The group initially wanted to compile to assembly, primarily for educational purposes, however as the initial brainstorming took place, the group quickly realized that they wanted a big set of possibilities, which would waste a lot of development time on code generation. 
+In order to save time, the group agreed on compiling to C++, the language that Arduino software is normally written in.
+When writing code in the Arduino IDE, it gives access to a couple of predefined methods for interacting with I/O devices and then through a C/C++ compiler\cite{ArFAQ}.
+However, the standard library is not accessible in the Arduino language, which means that dynamic lists, vectors, and other similar constructs are not accessible.
+Therefore it was required that the development team of HCL would implement some of these constructs manually, however this is still a lot easier than compiling to assembly directly. 
+The compilation process, with this is mind, will essentially be: Compile to the Arduino-subset of C++, and then let the Arduino software generate assembly code from the code generated by the HCL compiler.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: Perhaps add a diagram of this.

File: report/3.Theory/LanguageSpecification.tex [hunk 10]
```diff
@@ -126,6 +132,29 @@ \section{Language Specification}
 Not having operators for arithmetic operations also means that there are no native operator precedence in the language.
 This can be amended by structuring calculations in a left-to-right manner, or by encapsulating operations with parenthesis. 
 
+\textbf{High order functions}\\
+As mentioned in section \ref{subsec:hofhcl}, the language will include high-order functions. 
+To declare that a function will accept other functions, that the developer should simply declare a parameter of type "func" with the expected input and output types from the function.
+
+Both predeclared functions and anonymous in-line functions can be passed as parameters. 
+However, if using a predeclared function, the identifier should be prefixed with a colon to indicate that the function should be passed and not invoked.
```
Comment (nillas12) [inline report/3.Theory/LanguageSpecification.tex]: Change "However, if...." to "However, when...."

File: report/4.Solution/LanuageOfCompiler.tex (hunk 7 above)
Comment (Abildgaard96) [inline report/4.Solution/LanuageOfCompiler.tex]: rebuild sentence
A disadvantage could be that low-level languages might have faster execution time.
However, as HCL is built with the Arduino architecture in mind, it is not possible to create vast programs due to hardware limitations.
Therefore, the execution time of the compiler will never really affect the end user, as the difference should not be noticeable

File: report/4.Solution/LanuageOfCompiler.tex [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+\subsection{Output language of the compiler}
+The group initially wanted to compile to assembly, primarily for educational purposes, however as the initial brainstorming took place, the group quickly realized that they wanted a big set of possibilities, which would waste a lot of development time on code generation. 
+In order to save time, the group agreed on compiling to C++, the language that Arduino software is normally written in.
+When writing code in the Arduino IDE, it gives access to a couple of predefined methods for interacting with I/O devices and then through a C/C++ compiler\cite{ArFAQ}.
+However, the standard library is not accessible in the Arduino language, which means that dynamic lists, vectors, and other similar constructs are not accessible.
+Therefore it was required that the development team of HCL would implement some of these constructs manually, however this is still a lot easier than compiling to assembly directly.
```
Comment (Abildgaard96) [inline report/4.Solution/LanuageOfCompiler.tex]: Split sentence after manually. and start the next with "However, this...

File: report/4.Solution/LanuageOfCompiler.tex (hunk 6 above)
Comment (nillas12) [reply report/4.Solution/LanuageOfCompiler.tex]: Dog bliver det mange the i streg, dunno

File: report/4.Solution/LanuageOfCompiler.tex (hunk 7 above)
Comment (Abildgaard96) [reply report/4.Solution/LanuageOfCompiler.tex]: optionally, look at Casa's supercool comment with proposed reformulation already presented :heart:  :dagger:  :sunglasses: :beer: :hand: :namibia:

File: report/3.Theory/LanguageSpecification.tex [hunk 12]
```diff
@@ -49,21 +49,27 @@ \section{Language Specification}
...
 \begin{lstlisting}
-func<num,num,none> addNumbers = (num a, num b) -> none {
-	a + b print
+func[num,num,none] addNumbers = (num a, num b): num {
+	return a + b
 }
 \end{lstlisting}
 
+The return statement is noticeably missing in the first funciton.
+"return" is optional if the function body only contains a single expression.
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: expression should be "command"

File: report/3.Theory/LanguageSpecification.tex (hunk 2 above)
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: "that the developer should simply declare a parameter"

possibly "then the developer..."

as well as "with the expected input and output types from the function." -> "with the expected input and output types explicitly stated from the function." or something like that

File: report/4.Solution/LanuageOfCompiler.tex [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+However it is still possible to write methods that is utilizes a functional domain.
+
+This grants the developer a large amount of possibilities, while still being relatively easy to learn.
+A disadvantage could be that low-level languages might have faster executing time, however as HCL is built with the Arduino architecture in mind, it is not possible to create vast programs, as the architecture has some hardware limitations, so execution time of the compiler will never really affect the end user, as the difference should not be noticeable.
+
+\subsection{Output language of the compiler}
+The group initially wanted to compile to assembly, primarily for educational purposes, however as the initial brainstorming took place, the group quickly realized that they wanted a big set of possibilities, which would waste a lot of development time on code generation. 
+In order to save time, the group agreed on compiling to C++, the language that Arduino software is normally written in.
+When writing code in the Arduino IDE, it gives access to a couple of predefined methods for interacting with I/O devices and then through a C/C++ compiler\cite{ArFAQ}.
```
Comment (Lambcreek) [inline report/4.Solution/LanuageOfCompiler.tex]: This sentence doesn't make sense after "and"
I'm not quite sure what was meant to be written here.

File: report/3.Theory/LanguageSpecification.tex [hunk 14]
```diff
@@ -135,31 +164,26 @@ \section{Language Specification}
...
-\end{lstlisting}
-
-Obviously the sum function accepts a list that should be summed. 
-However, if the sum function could potentially also accept a modifier function, that were to alter the result of the sum function, it would be impossible for the compiler to determine whether print was that modifier function, or the sum should be passed as a parameter to the print function.
+Function overloading in HCL is limited in that functions can not be overloaded with same parameters but different return types. 
+Functions can only be overloaded to return different types, if the types of the parameters vary.
+Furthermore overloading with different amounts of parameters are also not possible in HCL.
+This is because of multiple reasons, one being simplicity for the user, not accidentally adding an unexpected parameter to a function call. 
+Another being simplicity in the parsing of the function parameters. 
+Since the ability to overload with different amount of parameters brings next to nothing to the language, it was decided that it was not worth the effort to implement.
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: write something on the lines of that functions can have varying length of parameters by usage of tuples or lists.

For instance if i want to either use 2 or 3 elements i could make a function like: "(num x, num y): num" AND another one like "(num x, tuple<num,num>): num".. which also adds required brackets to make it more easily readable for some cases etc etc.

**Pull request merged**